web/src/copy_messages.ts

```diff
@@ -119,54 +119,88 @@ function get_nearest_html_element(node: Node | null): Element | null {
+        }
+    } else {
+        if (is_range_start) {
+            range.setStart(selection_element.closest(".katex")!, 0);
+        } else {
+            // The offset 2 selects the two children of `.katex`
+            // namely `.katex-mathml` and `.katex-html`
+            range.setEnd(selection_element.closest(".katex")!, 2);
+        }
+    }
+}
 /*
     This is done to make the copying within math blocks smarter.
     We mutate the selection for selecting singular katex expressions
     within math blocks when applicable, which will be
     converted into the inline $$<expr>$$ syntax.
 
-    In case when a single expression or its subset is selected
-    within a math block, we adjust the selection so that it
-    selects the katex span which is parenting that expression.
-    This converts the selection into an inline expression as
-    per the turndown rules below.
-
-    We want to avoid this behavior if the selection
-    spreads across multiple katex displays i.e. the
-    focus and anchor are not part of the same katex span.
+    We expand the selection range only in the following cases:
+    1. Either the anchor or focus or both nodes are within an inline expression where the selection covers
+    one or more math expressions.
+    2. Either the anchor, focus, or both nodes are within a math block where the selection covers one or more
+    math expressions.
+
+    In principle, we only need to expand the start of the selection range for the
+    cases where multiple expressions are selected because the end of the range
+    always contains the annotation element in case it lies within the math block.
+
+    But, we still expand the end of the range to select the complete expression.
+    This is done in cases where the anchor/focus denoting the end of the range
+    lies inside the math block or expression(depending on whether anchor comes after the focus or
+    vice-versa in the DOM) as part of the effort to keep selection and paste content
+    as close as possible, to eventually tackle #6316.
+
+    The primary objective of this function is to include the `<annotation>` tag within
+    the selection range so that it can be utilized later by `paste_handler_converter`
+    as `paste_html`.
 */
-function improve_katex_selection_range(selection: Selection): void {
-    const anchor_element = get_nearest_html_element(selection.anchorNode);
-    const focus_element = get_nearest_html_element(selection.focusNode);
-    // If the anchor and focus end up in different katex spans, this selection
-    // isn't meant to be an inline expression, so we perform an early return.
-    if (
-        focus_element &&
-        anchor_element &&
-        focus_element?.closest(".katex") !== anchor_element?.closest(".katex")
-    ) {
+function improve_katex_selection_range(range: Range): void {
+    const anchor_element = get_nearest_html_element(range.startContainer);
+    const focus_element = get_nearest_html_element(range.endContainer);
```

Comment:
I think you want to name these `start_element` and `end_element`, since they no longer are anchor and focus.